File: scripts/generate_config_index.py
#!/usr/bin/env python3
"""
Generate a comprehensive index of all application configuration locations
from Mackup's application database.

This script extracts all config file paths from the .cfg files and outputs
them in JSON and YAML formats for use with custom sync engines.
"""

import configparser
import json
import os
from pathlib import Path


def parse_app_config(cfg_path: str) -> dict:
    """Parse a single application .cfg file."""
    config = configparser.ConfigParser(allow_no_value=True)
    config.optionxform = str  # Preserve case

    config.read(cfg_path)

    app_data = {
        "name": config.get("application", "name", fallback="Unknown"),
        "config_file": os.path.basename(cfg_path),
        "configuration_files": [],
        "xdg_configuration_files": [],
    }

    if config.has_section("configuration_files"):
        app_data["configuration_files"] = list(config.options("configuration_files"))

    if config.has_section("xdg_configuration_files"):
        app_data["xdg_configuration_files"] = list(
            config.options("xdg_configuration_files")
        )

    return app_data


def generate_index():
    """Generate the complete configuration index."""
    # Find the applications directory
    script_dir = Path(__file__).parent
    apps_dir = script_dir.parent / "mackup" / "applications"

    if not apps_dir.exists():
        raise FileNotFoundError(f"Applications directory not found: {apps_dir}")

    applications = {}
    all_paths = {
        "home_relative": set(),  # Paths relative to ~
        "xdg_relative": set(),  # Paths relative to $XDG_CONFIG_HOME
        "macos_library": set(),  # macOS ~/Library paths
    }

    # Parse all .cfg files
    for cfg_file in sorted(apps_dir.glob("*.cfg")):
        app_id = cfg_file.stem
        app_data = parse_app_config(str(cfg_file))
        applications[app_id] = app_data

        # Categorize paths
        for path in app_data["configuration_files"]:
            all_paths["home_relative"].add(path)
            if path.startswith("Library/"):
                all_paths["macos_library"].add(path)

        for path in app_data["xdg_configuration_files"]:
            all_paths["xdg_relative"].add(path)

    # Build the complete index
    index = {
        "metadata": {
            "source": "Mackup Application Database",
            "source_url": "https://github.com/lra/mackup",
            "total_applications": len(applications),
            "total_home_paths": len(all_paths["home_relative"]),
            "total_xdg_paths": len(all_paths["xdg_relative"]),
            "total_macos_library_paths": len(all_paths["macos_library"]),
            "description": "Configuration file locations for application settings backup/sync",
        },
        "path_resolution": {
            "home_relative": "Paths are relative to $HOME (e.g., ~/.bashrc)",
            "xdg_relative": "Paths are relative to $XDG_CONFIG_HOME (defaults to ~/.config)",
            "macos_library": "Paths starting with Library/ are macOS-specific (~/Library/...)",
        },
        "applications": applications,
        "all_paths": {
            "home_relative": sorted(all_paths["home_relative"]),
            "xdg_relative": sorted(all_paths["xdg_relative"]),
            "macos_library_only": sorted(all_paths["macos_library"]),
        },
    }

    return index


def main():
    index = generate_index()

    output_dir = Path(__file__).parent.parent / "config_index"
    output_dir.mkdir(exist_ok=True)

    # Write JSON
    json_path = output_dir / "application_configs.json"
    with open(json_path, "w") as f:
        json.dump(index, f, indent=2)
    print(f"Generated: {json_path}")

    # Write YAML
    yaml_path = output_dir / "application_configs.yaml"
    with open(yaml_path, "w") as f:
        f.write("# Application Configuration Index\n")
        f.write("# Generated from Mackup application database\n")
        f.write("# For use with custom sync engines\n\n")
        f.write(f"metadata:\n")
        for key, value in index["metadata"].items():
            f.write(f"  {key}: {json.dumps(value)}\n")
        f.write("\npath_resolution:\n")
        for key, value in index["path_resolution"].items():
            f.write(f"  {key}: {json.dumps(value)}\n")
        f.write("\napplications:\n")
        for app_id, app_data in sorted(index["applications"].items()):
            f.write(f"  {app_id}:\n")
            f.write(f"    name: {json.dumps(app_data['name'])}\n")
            if app_data["configuration_files"]:
                f.write(f"    configuration_files:\n")
                for path in app_data["configuration_files"]:
                    f.write(f"      - {json.dumps(path)}\n")
            if app_data["xdg_configuration_files"]:
                f.write(f"    xdg_configuration_files:\n")
                for path in app_data["xdg_configuration_files"]:
                    f.write(f"      - {json.dumps(path)}\n")
    print(f"Generated: {yaml_path}")

    # Write a simple flat list for quick reference
    flat_path = output_dir / "all_paths.txt"
    with open(flat_path, "w") as f:
        f.write("# All Configuration Paths\n")
        f.write("# Format: [APP_ID] PATH\n\n")
        f.write("## Home-relative paths (relative to ~)\n\n")
        for app_id, app_data in sorted(index["applications"].items()):
            for path in sorted(app_data["configuration_files"]):
                f.write(f"[{app_id}] ~/{path}\n")
        f.write("\n## XDG paths (relative to $XDG_CONFIG_HOME, default ~/.config)\n\n")
        for app_id, app_data in sorted(index["applications"].items()):
            for path in sorted(app_data["xdg_configuration_files"]):
                f.write(f"[{app_id}] $XDG_CONFIG_HOME/{path}\n")
    print(f"Generated: {flat_path}")

    # Print summary
    print(f"\n=== Summary ===")
    print(f"Total applications: {index['metadata']['total_applications']}")
    print(f"Home-relative paths: {index['metadata']['total_home_paths']}")
    print(f"XDG paths: {index['metadata']['total_xdg_paths']}")
    print(f"macOS Library paths: {index['metadata']['total_macos_library_paths']}")


if __name__ == "__main__":
    main()

File: scripts/generate-config-index.ts
#!/usr/bin/env bun
/**
 * Generate a comprehensive index of all application configuration locations
 * from Mackup's application database.
 *
 * This script extracts all config file paths from the .cfg files and outputs
 * them in JSON and YAML formats for use with custom sync engines.
 */

import { readdir } from 'node:fs/promises';
import { join, basename } from 'node:path';
import { parseMackupConfig, type MackupAppConfig } from './lib/ini-parser';

interface ApplicationIndex {
  [appId: string]: {
    name: string;
    config_file: string;
    configuration_files: string[];
    xdg_configuration_files: string[];
  };
}

interface AllPaths {
  home_relative: string[];
  xdg_relative: string[];
  macos_library_only: string[];
}

interface ConfigIndex {
  metadata: {
    source: string;
    source_url: string;
    total_applications: number;
    total_home_paths: number;
    total_xdg_paths: number;
    total_macos_library_paths: number;
    description: string;
  };
  path_resolution: {
    home_relative: string;
    xdg_relative: string;
    macos_library: string;
  };
  applications: ApplicationIndex;
  all_paths: AllPaths;
}

async function parseAllConfigs(appsDir: string): Promise<Map<string, MackupAppConfig>> {
  const configs = new Map<string, MackupAppConfig>();

  const files = await readdir(appsDir);
  const cfgFiles = files.filter((f) => f.endsWith('.cfg')).sort();

  for (const cfgFile of cfgFiles) {
    const filePath = join(appsDir, cfgFile);
    const content = await Bun.file(filePath).text();
    const appId = basename(cfgFile, '.cfg');
    const config = parseMackupConfig(content, cfgFile);
    configs.set(appId, config);
  }

  return configs;
}

function buildIndex(configs: Map<string, MackupAppConfig>): ConfigIndex {
  const applications: ApplicationIndex = {};
  const homeRelativePaths = new Set<string>();
  const xdgRelativePaths = new Set<string>();
  const macosLibraryPaths = new Set<string>();

  for (const [appId, config] of configs) {
    applications[appId] = {
      name: config.name,
      config_file: config.configFile,
      configuration_files: config.configurationFiles,
      xdg_configuration_files: config.xdgConfigurationFiles,
    };

    for (const path of config.configurationFiles) {
      homeRelativePaths.add(path);
      if (path.startsWith('Library/')) {
        macosLibraryPaths.add(path);
      }
    }

    for (const path of config.xdgConfigurationFiles) {
      xdgRelativePaths.add(path);
    }
  }

  return {
    metadata: {
      source: 'Mackup Application Database',
      source_url: 'https://github.com/lra/mackup',
      total_applications: configs.size,
      total_home_paths: homeRelativePaths.size,
      total_xdg_paths: xdgRelativePaths.size,
      total_macos_library_paths: macosLibraryPaths.size,
      description: 'Configuration file locations for application settings backup/sync',
    },
    path_resolution: {
      home_relative: 'Paths are relative to $HOME (e.g., ~/.bashrc)',
      xdg_relative: 'Paths are relative to $XDG_CONFIG_HOME (defaults to ~/.config)',
      macos_library: 'Paths starting with Library/ are macOS-specific (~/Library/...)',
    },
    applications,
    all_paths: {
      home_relative: [...homeRelativePaths].sort(),
      xdg_relative: [...xdgRelativePaths].sort(),
      macos_library_only: [...macosLibraryPaths].sort(),
    },
  };
}

function generateYaml(index: ConfigIndex): string {
  const lines: string[] = [
    '# Application Configuration Index',
    '# Generated from Mackup application database',
    '# For use with custom sync engines',
    '',
    'metadata:',
  ];

  for (const [key, value] of Object.entries(index.metadata)) {
    lines.push(`  ${key}: ${JSON.stringify(value)}`);
  }

  lines.push('', 'path_resolution:');
  for (const [key, value] of Object.entries(index.path_resolution)) {
    lines.push(`  ${key}: ${JSON.stringify(value)}`);
  }

  lines.push('', 'applications:');
  const sortedApps = Object.entries(index.applications).sort(([a], [b]) => a.localeCompare(b));

  for (const [appId, appData] of sortedApps) {
    lines.push(`  ${appId}:`);
    lines.push(`    name: ${JSON.stringify(appData.name)}`);

    if (appData.configuration_files.length > 0) {
      lines.push('    configuration_files:');
      for (const path of appData.configuration_files) {
        lines.push(`      - ${JSON.stringify(path)}`);
      }
    }

    if (appData.xdg_configuration_files.length > 0) {
      lines.push('    xdg_configuration_files:');
      for (const path of appData.xdg_configuration_files) {
        lines.push(`      - ${JSON.stringify(path)}`);
      }
    }
  }

  return lines.join('\n');
}

function generateFlatList(index: ConfigIndex): string {
  const lines: string[] = [
    '# All Configuration Paths',
    '# Format: [APP_ID] PATH',
    '',
    '## Home-relative paths (relative to ~)',
    '',
  ];

  const sortedApps = Object.entries(index.applications).sort(([a], [b]) => a.localeCompare(b));

  for (const [appId, appData] of sortedApps) {
    for (const path of appData.configuration_files.sort()) {
      lines.push(`[${appId}] ~/${path}`);
    }
  }

  lines.push('', '## XDG paths (relative to $XDG_CONFIG_HOME, default ~/.config)', '');

  for (const [appId, appData] of sortedApps) {
    for (const path of appData.xdg_configuration_files.sort()) {
      lines.push(`[${appId}] $XDG_CONFIG_HOME/${path}`);
    }
  }

  return lines.join('\n');
}

async function main() {
  const scriptDir = import.meta.dir;
  const projectRoot = join(scriptDir, '..');
  const appsDir = join(projectRoot, 'mackup', 'applications');
  const outputDir = join(projectRoot, 'config_index');

  console.log('Parsing application configs...');
  const configs = await parseAllConfigs(appsDir);

  console.log('Building index...');
  const index = buildIndex(configs);

  // Ensure output directory exists
  await Bun.write(join(outputDir, '.gitkeep'), '');

  // Write JSON
  const jsonPath = join(outputDir, 'application_configs.json');
  await Bun.write(jsonPath, JSON.stringify(index, null, 2));
  console.log(`Generated: ${jsonPath}`);

  // Write YAML
  const yamlPath = join(outputDir, 'application_configs.yaml');
  await Bun.write(yamlPath, generateYaml(index));
  console.log(`Generated: ${yamlPath}`);

  // Write flat list
  const flatPath = join(outputDir, 'all_paths.txt');
  await Bun.write(flatPath, generateFlatList(index));
  console.log(`Generated: ${flatPath}`);

  // Print summary
  console.log('\n=== Summary ===');
  console.log(`Total applications: ${index.metadata.total_applications}`);
  console.log(`Home-relative paths: ${index.metadata.total_home_paths}`);
  console.log(`XDG paths: ${index.metadata.total_xdg_paths}`);
  console.log(`macOS Library paths: ${index.metadata.total_macos_library_paths}`);
}

main().catch((err) => {
  console.error('Error:', err);
  process.exit(1);
});
